Crash on opening .tgz file

## Issue explanation (write below this line)

Tried to open .tgz file and got this crash message. App is up to date and already tried force stopping and clearing cache.

## Exception
* __App Name:__ Amaze File Manager
* __Package:__ com.amaze.filemanager
* __Version:__ 3.6.1
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 11 - 30
* __Device:__ sargo
* __Model:__ Pixel 3a
* __Product:__ sargo
<details><summary><b>Crash log </b></summary><p>

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
Caused by: java.lang.reflect.InvocationTargetException
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
	... 1 more
Caused by: java.lang.NoSuchMethodException: org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream.<init> [class java.io.InputStream]
	at java.lang.Class.getConstructor0(Class.java:2332)
	at java.lang.Class.getDeclaredConstructor(Class.java:2170)
	at com.amaze.filemanager.asynchronous.asynctasks.compress.AbstractCompressedTarArchiveHelperTask.<init>(AbstractCompressedTarArchiveHelperTask.kt:46)
	at com.amaze.filemanager.asynchronous.asynctasks.compress.GzipHelperTask.<init>(GzipHelperTask.kt:38)
	at com.amaze.filemanager.filesystem.compressed.showcontents.helpers.GzipDecompressor.changePath(GzipDecompressor.kt:37)
	at com.amaze.filemanager.filesystem.compressed.showcontents.helpers.GzipDecompressor.changePath(GzipDecompressor.kt:31)
	at com.amaze.filemanager.ui.fragments.CompressedExplorerFragment.changePath(CompressedExplorerFragment.java:435)
	at com.amaze.filemanager.ui.fragments.CompressedExplorerFragment.onActivityCreated(CompressedExplorerFragment.java:226)
	at androidx.fragment.app.Fragment.performActivityCreated(Fragment.java:2983)
	at androidx.fragment.app.FragmentStateManager.activityCreated(FragmentStateManager.java:577)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:285)
	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:2177)
	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:2088)
	at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1990)
	at androidx.fragment.app.FragmentManager$5.run(FragmentManager.java:524)
	at android.os.Handler.handleCallback(Handler.java:938)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:223)
	at android.app.ActivityThread.main(ActivityThread.java:7660)
	... 3 more

```
</details>
<hr>

--- a/app/proguard.cfg
+++ b/app/proguard.cfg
@@ -118,3 +118,9 @@
 #Ignore test classes see tests-proguard.cfg
 -dontwarn android.test.**
 -dontwarn org.junit.**
+
+#Commons-compress. See #2647.
+#Keep constructors that involves an InputStream
+-keepclassmembers class * extends org.apache.commons.compress.compressors.CompressorInputStream {
+  <init>(java.io.InputStream);
+}

--- a/app/src/main/java/com/amaze/filemanager/filesystem/compressed/Extensions.kt
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/compressed/Extensions.kt
@@ -1,0 +1,43 @@
+/*
+ * Copyright (C) 2014-2021 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
+ * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
+ *
+ * This file is part of Amaze File Manager.
+ *
+ * Amaze File Manager is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.amaze.filemanager.filesystem.compressed
+
+import android.os.Build.VERSION.SDK_INT
+import android.os.Build.VERSION_CODES.M
+import com.github.junrar.Archive
+
+/**
+ * Extension function to patch [Archive.isPasswordProtected] which uses API that is not available
+ * for Android 6.0 or lower.
+ *
+ * @see [Archive.isPasswordProtected]
+ * @see [java.util.stream.Stream]
+ * @return true if archive is password protected
+ */
+fun Archive.isPasswordProtectedCompat(): Boolean {
+    return if (SDK_INT > M) {
+        this.isPasswordProtected
+    } else {
+        fileHeaders.any { obj ->
+            obj.isEncrypted
+        }
+    }
+}

--- a/app/src/main/java/com/amaze/filemanager/filesystem/compressed/extractcontents/helpers/RarExtractor.kt
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/compressed/extractcontents/helpers/RarExtractor.kt
@@ -28,6 +28,7 @@ import com.amaze.filemanager.filesystem.FileUtil
 import com.amaze.filemanager.filesystem.MakeDirectoryOperation
 import com.amaze.filemanager.filesystem.compressed.CompressedHelper
 import com.amaze.filemanager.filesystem.compressed.extractcontents.Extractor
+import com.amaze.filemanager.filesystem.compressed.isPasswordProtectedCompat
 import com.amaze.filemanager.filesystem.files.GenericCopyUtil
 import com.github.junrar.Archive
 import com.github.junrar.exception.CorruptHeaderException
@@ -68,7 +69,7 @@ class RarExtractor(
                 }
             }.getOrNull()!!
 
-            if (rarFile.isPasswordProtected || rarFile.isEncrypted) {
+            if (rarFile.isPasswordProtectedCompat() || rarFile.isEncrypted) {
                 if (ArchivePasswordCache.getInstance().containsKey(filePath)) {
                     runCatching {
                         tryExtractSmallestFileInArchive(context, rarFile)

